autogen/agentchat/contrib/capabilities/generate_images.py

```diff
@@ -0,0 +1,156 @@
+        img_data = img_utils.get_image_data(image_url)
+        cache[key] = img_data
+        return img_utils._to_pil(img_data)
+
+
+class ImageGeneration(AgentCapability):
+    def __init__(
+        self, image_generator: ImageGenerator, text_analyzer_llm_config: Optional[Dict] = None, verbosity: int = 0
+    ):
+        self._image_generator = image_generator
+        self._text_analyzer_llm_config = text_analyzer_llm_config
+        self._verbosity = verbosity
+
+        self._agent: Optional[ConversableAgent] = None
+        self._text_analyzer: Optional[TextAnalyzerAgent] = None
+
+    def add_to_agent(self, agent: ConversableAgent):
+        self._agent = agent
+
+        agent.register_reply([Agent, None], self._image_gen_reply, position=2)
+
+        self._text_analyzer_llm_config = self._text_analyzer_llm_config or agent.llm_config
+        self._text_analyzer = TextAnalyzerAgent(llm_config=self._text_analyzer_llm_config)
+
+        agent.update_system_message(agent.system_message + "\n" + SYSTEM_MESSAGE)
+        # Hack to update the description of the agent, useful when used in a group chat
+        agent.description += "\n" + DESCRIPTION_MESSAGE
+
+    def _image_gen_reply(
+        self,
+        reciepient: ConversableAgent,
+        messages: Optional[List[Dict]],
+        sender: Optional[Agent] = None,
+        config: Optional[Any] = None,
+    ) -> Tuple[bool, Union[str, Dict, None]]:
+        if messages is None:
+            return False, None
+
+        last_message = messages[-1]["content"]
+        if isinstance(last_message, dict):
+            last_message = last_message.get("text", "")
+
+        if not last_message:
+            return False, None
+
+        if self._should_generate_image(last_message):
+            prompt = self._analyze_text(
+                last_message,
+                "In detail, please provide the prompt to generate the image described in the TEXT. DO NOT include any advice.",
+            )
+            image = self._image_generator.generate_image(prompt)
+
+            # TODO: DELETE THIS AFTER TESTING
+            image.show()
+
+            return True, {
+                "content": [
+                    {"type": "text", "text": f"Generated an image with the prompt: {prompt}"},
+                    {"type": "image_url", "image_url": {"url": image}},
+                ]
+            }
+        else:
+            return False, None
+
+    def _should_generate_image(self, message: str) -> bool:
+        response = self._analyze_text(
+            message,
+            "Does any part of the TEXT ask the agent to generate or modify an image? Answer with just one word, yes or no.",
+        )
+        return "yes" in response.lower()
+
+    def _analyze_text(self, text_to_analyze: str, analysis_instructions: str) -> str:
+        assert self._text_analyzer is not None
+        assert self._agent is not None
+
+        self._text_analyzer.reset()
+        self._agent.send(
```

Comment:
@rickyloynd-microsoft  I know splitting the message into two messages is from the "teachability". It can make LLM's answer more accurate (aka, following the format). 

However, some smaller models would not accept two consecutive messages from the same agent. Let's find out some other ways in the future. For here, I am good with it.

Reply:
The LLM doesn't see two messages from the same agent. TextAnalyzerAgent combines them into one before sending it to the LLM.